Feature/cogs

MonkfishFillet opened the pull request:
Refactored two of the bot functions into cogs and allows us to:

- Reduces the size of robbot.py
- Add and remove features using the Cog commands while Robbot is running
- Organize this shit just a little bit more

Also tried to condense all of the helper classes into a single Util module - May come back to break this up eventually

MonkfishFillet committed:
Created Configuration Abstract Class

MonkfishFillet committed:
Added venv+jupyterlab .gitignore Entries

MonkfishFillet committed:
User Colors Added to Colors Module

MonkfishFillet committed:
Pin & Pin Embedder Fix

MonkfishFillet committed:
Class Comment Overhaul

MonkfishFillet committed:
bot.command Name/Alias Update

MonkfishFillet committed:
Utilility Consolidation + Cogs Added

MonkfishFillet committed:
Fixed Rolls File / Added Requirments.txt

MonkfishFillet committed:
Moved Requirements.txt

MonkfishFillet committed:
Alpha Music Theme Player Created - General Cleanup

MonkfishFillet committed:
Added Emojis - Cum Listener

MonkfishFillet committed:
Sanitizer Service Created

MonkfishFillet committed:
Embedder Service Created

MonkfishFillet committed:
Embedder Service Fixes

MonkfishFillet committed:
Removed Formatter

MonkfishFillet committed:
Bartender Cog Updated w/ Services

MonkfishFillet committed:
Pinner Cog Updated w/ Services

MonkfishFillet committed:
Requester Service Implemented"

MonkfishFillet committed:
Removed Cringe Comments

DonChalupa commented on robbot.py:
```diff
@@ -26,6 +16,33 @@
+'''
+def load_cogs():
+    for files in os.listdir('./cogs'):
+        if files.endswith('.py'):              
```
Since `files` refers to a single element in the list of filenames returned by `os.listdir()`, not a list itself, the name `file` would be more precise.

```suggestion
    for file in os.listdir('./cogs'):
        if file.endswith('.py'):              
```

DonChalupa commented on robbot.py:
```diff
@@ -26,6 +16,33 @@
+    for files in os.listdir('./cogs'):
+        if files.endswith('.py'):              
+            bot.load_extension(f'cogs.{files[:-3]}')
+            print(f'{files[:-3]} extension loaded')
```
A more readable way of trimming the file extension would be to call `os.path.splitext(file)` and use that result in the load call. 

```suggestion
            cog_name, file_ext = os.path.splitext(file)
            bot.load_extension(f'cogs.{cog_name}')
            print(f'{cog_name} extension loaded')
```

Optionally, `file_ext` could be named `_`, which by convention denotes that it's unused - personally feel that `file_ext` is a bit more clear.

DonChalupa commented on robbot.py:
```diff
@@ -26,6 +16,33 @@
+    for files in os.listdir('./cogs'):
+        if files.endswith('.py'):              
+            bot.unload_extension(f'cogs.{files[:-3]}')
+            print(f'{files[:-3]} extension unloaded')
```
See comments on `load_cogs()`.

```suggestion
    for file in os.listdir('./cogs'):
        if file.endswith('.py'):              
            cog_name, file_ext = os.path.splitext(file)
            bot.unload_extension(f'cogs.{cog_name}')
            print(f'{cog_name} extension unloaded')
```

DonChalupa commented on cogs/Bartender.py:
```diff
@@ -0,0 +1,26 @@
+        self._sanitizer = Sanitizer.SantizerFactory.get_sanitizer(Bartender.factory_type)
+        self._embedder = Embedder.EmbedderFactory.get_embedder(Bartender.factory_type)
+    
+    @commands.command(name='drink', description='Get a nice, freshing drink from the Bartender',aliases=['drinks', 'cocktail', 'cocktails'])
```
Possible typo

```suggestion
    @commands.command(name='drink', description='Get a nice, refreshing drink from the Bartender',aliases=['drinks', 'cocktail', 'cocktails'])
```

DonChalupa commented on cogs/Music.py:
```diff
@@ -0,0 +1,62 @@
+        if(len(args) == 1):
+            url = self._url_factory(args[0])
+
+        if(url is None):
+            await ctx.channel.send("That's not a valid music theme.")
+
+        else:
+            await ctx.channel.send(f"Here you are, Bud:\n!play {url}")
+    
+    def _url_factory(self, theme):
+        print(theme)
+        if(theme == 'session-start'):
+            return self._links['session-start']
+
+        elif(theme == 'battle-music'):
+            return self._links['battle-music'][0]
+
+        elif(theme == 'desert'):
+            return self._links['desert'][0]
+
+        elif(theme == 'town-center'):
+            return self._links['town-center'][0]
+
+        elif(theme == 'bar'):
+            return self._links['bar'][0]
+
+        elif(theme == 'docks'):
+            return self._links['docks'][0]
+
+        elif(theme == 'cave'):
+            return self._links['cave'][0]
+    
+        else:
+            return None
+
+
+def setup(bot):
+    bot.add_cog(Music(bot))
+
+
```
While this cog doesn't necessarily conflict with the `Music` module on my branch, it does depend on Hydra (which we no longer have in the discord) and essentially duplicates the functionality found in the `Music` module.

Rather than include this cog, we can expand on the existing `Music` module to include the functionality it provides, i.e. providing aliases for Youtube URLs. This is an issue we should hold off on until at least this merge is finished.

```suggestion

```

MonkfishFillet replied:
Outdated and cringe cog, removed

DonChalupa commented on config.json:
```diff
@@ -1,11 +1,41 @@
         "Cannot predict now.",
         "Concentrate and ask again.",
-        "Don't count on it.",
+        "Dont count on it.",
```
```suggestion
        "Don't count on it.",
```

DonChalupa commented on util/Utility.py:
```diff
@@ -0,0 +1,190 @@
+    ORLANDO = 324694553200427009
+    SPEED = 324694586855391244
+    SUPERNUT = 324694596397432833
+    TIFA = 324694604903481344
+    VINCENT = 324694614823010314
+    BUTTHETEACHINGS = 324694624708984832
+    ZILTOID = 324694636972998657
+    AUTISTICSCREECHING = 324694662759841792
+    DERICK = 324695421639196672
+    HURR = 324696431904686090
+    CRAWLING = 324699598914453504
+    THATSRAD = 324706382781743105
+    STUPID_IDIOT = 324724786557353984
+    SMUGTEACHER = 324745655878811648
+    DDOG = 324750990630846474
+    SAKEGREMLIN = 324752939904270336
+    Q_ = 325104825882902538
+    DOTHEYSMUGEVERYTIME = 325105720121098240
+    HAHAA = 325106039345381376
+    KIRYU = 325106393701023744
+    LOLOL = 325107124931788800
+    LETSNOTDOTHATTODAY = 336164623453650944
+    RIBBERTSTIBBERT = 399288392321335307
+    HMMM = 399302648169496577
+    HMMMMM = 399304587745689600
+    YOTSUBATUNES = 399311065416204288
+    NOWIMMOTIVATED = 458107334926008331
+    BIGFOOT = 468234675622641674
+    BIGBOSSU = 468564992334233600
+    OHJEEZ = 543988083306266624
+    PERSONA = 543996812160401408
+    SMIRK = 734835488166379550
+    HMM = 791116665202016256
+    UM = 791118149063147560
+    EDGYSMILE = 796455827556270120
+    CRINGESHOCK = 796455847127416882
+    HEH = 803777598877597736
+    DROOL = 804524633998622740
+    SAYHUH = 832708202852253706
+#endregion
```
There's no need to hardcode all of this information because we should be able to get it from calls to `guild.emojis`, `guild.members`, `guild.roles`, `guild.channels`, the `Colour` class, etc. Additionally, all of this information is subject to change, duplicating the info here means any changes to the server will require changes to this file.

Whenever this information is needed, we should get it dynamically whenever possible. There are situations where we may want to get a specific channel or user by name, and in that case the duplication of info is unavoidable (e.g., a hypothetical `foo()` command that posts a message in the "#foo" channel will have to know what channel to look for, so the channel name "#foo" will have to be stored in code or config somewhere in the process), but in other situations a lookup will avoid code changes when the server changes.

For example, if we want a user's color for embeds, instead of hardcoding their color here, we can lookup the user's roles, determine their primary role, and then get the role's color. There may be times we want to get a user by their actual identity (e.g., we want a listener that responds only to James' messages), in which case hardcoding the user ID is pretty unreadable, so a simple map of colloquial names to user IDs would be acceptable and sufficient for those cases - somewhat like the `config.json` dictionary, but just first names to IDs.

MonkfishFillet replied:
I agree with majority of this - hard coding this isn't a good look for a scalability. 

User colors was something I wanted to be customizable for each user and not dependent of role, though it's super low priority right now and can be on the back burner. Burning this file to the ground.

DonChalupa commented on config.json:
```diff
@@ -1,11 +1,41 @@
 {
+    "users": {
+        "178851504076095488": {
+            "name": "Brady",
+            "embed_color": "15844367"
+        },
+        "323968599687430145": {
+            "name":  "Carlie",
+            "embed_color": "10181046"
+        },
+        "191384926321508353": {
+            "name": "James",
+            "embed_color": "9807270"
+        },
+        "179051738354024449": {
+            "name": "Derick",
+            "embed_color": "3447003"
+        },
+        "636724139922554893": {
+            "name": "Keeth",
+            "embed_color": "15158332"
+        },
+        "209101464944115713": {
+            "name": "Rob",
+            "embed_color": "15105570"
+        },
+        "422470851435298818": {
+            "name": "Robbot",
+            "embed_color": "16580705"
+        }
+    },
```
See comments on `util/Utility.py`.

Embed colors shouldn't be hardcoded here as they are subject to change (as long as we're basing them off of roles, anyway).

A dictionary mapping colloquial names to user IDs makes sense for referring to specific users in code in a readable fashion - use the colloquial name to find the user ID from this dict, then use the ID to get the rest of the user's info.

DonChalupa commented on config.json:
```diff
@@ -20,5 +50,30 @@
         "Yes - definitely.",
         "You may rely on it."
-    ]
+    ],
+    "colors":{
+        "DEFAULT": 0,
+        "AQUA": 1752220,
+        "GREEN": 3066993,
+        "BLUE": 3447003,
+        "PURPLE": 10181046,
+        "GOLD": 15844367,
+        "ORANGE": 15105570,
+        "RED": 15158332,
+        "GREY": 9807270,
+        "DARKER_GREY": 8359053,
+        "NAVY": 3426654,
+        "DARK_AQUA": 1146986,
+        "DARK_GREEN": 2067276,
+        "DARK_BLUE": 2123412,
+        "DARK_PURPLE": 7419530,
+        "DARK_GOLD": 12745742,
+        "DARK_ORANGE": 11027200,
+        "DARK_RED": 10038562,
+        "DARK_GREY": 9936031,
+        "LIGHT_GREY": 12370112,
+        "DARK_NAVY": 2899536,
+        "LUMINOUS_VIVID_PINK": 16580705,
+        "DARK_VIVID_PINK": 12320855
+    }
```
All of this info is in the API's `Colour` class. There should be no reason to duplicate it here.

```suggestion
    ]
```

DonChalupa commented on robbot.py:
```diff
@@ -26,6 +16,33 @@
+            bot.unload_extension(f'cogs.{files[:-3]}')
+            print(f'{files[:-3]} extension unloaded')
+
+@bot.command(name='refresh', description='Reloads all cogs', aliases=['reload'], hidden=True)
+async def refresh(ctx: commands.Context):
+    print('Reloading Cogs...\n')
+    unload_cogs()
+    load_cogs()
+    print('Cogs reloaded successfully')
+    reaction = '👍'
+    await ctx.message.add_reaction(reaction)
```
I'm not sure about this yet, but I think this might be antithetical to the purpose of cogs & extensions. If I understand them right, the idea is to allow reloading of individual cogs at runtime without affecting the rest of the cogs, e.g. `bot.reload_extension('Music')` rather than unloading and loading all of them.

See [this example](https://github.com/Rapptz/RoboDanny/blob/master/cogs/admin.py#L18) I found that I think demonstrates this point better than I can explain. Also note that he has the commands in an `Admin` cog, probably a good idea for us as well.

DonChalupa reviewed (changes requested):
I still have to dig in to the `models/*`, `services/*`, etc. semantics to understand how that all works together, but here's what I'm thinking with what I've gone through so far.

MonkfishFillet committed:
Cog Function Refactor

MonkfishFillet committed:
Typo Fix

MonkfishFillet committed:
Removal of Music Cog

MonkfishFillet committed:
Removal of Utility File

DonChalupa commented:
Also, please pull the latest develop branch, merge it into this feature branch, resolve conflicts, then push those fixes on this feature branch to remove conflicts.

MonkfishFillet committed:
Removed Cum

MonkfishFillet committed:
Conflict Resolution

MonkfishFillet committed:
Merge branch 'dev-merge' into hot-merge

MonkfishFillet committed:
Sloppy commit

DonChalupa commented on cogs/Pinner.py:
```diff
@@ -0,0 +1,37 @@
+from discord.ext import commands
+from models.Pin import Pin
+from services import Embedder
+from util.Utility import Channels
```
`util.Utililty` no longer exists per latest commits

```suggestion
```

MonkfishFillet committed:
Typo Change

MonkfishFillet reviewed (commented):
Should be good to go, finally on a clean slate. Let's try to get this merged before the new year.

DonChalupa commented on cogs/Pinner.py:
```diff
@@ -0,0 +1,37 @@
+                await ctx.message.channel.send('You have to reply .pin to the message you want pinned.')
+                return
+            
+            pin_channel = self.bot.get_channel(Channels.PINS.value)
```
Broken by deletion of `util.Utility`. Honestly not sure the best way to handle this, seems like there's no way around hardcoding the channel name or ID somewhere.

You could search the guild's channel list for a channel called "pins" (e.g. `channel = next(filter(lambda x: x.name == "pins", ctx.guild.channels), None)`), but then if the channel name changed (for whatever reason) you'd have to change this line. You could create a constant `PIN_CHANNEL_NAME = "pins"` either in this file or in config (which is basically what you had before). That would be better since in the event of a change, you'd only have to change it in one spot in the code.

You could search for it by the channel ID instead of the name, so the channel name can change without breaking the code, but if the channel were deleted and re-added for some reason it would still break.

I think the only way we could do it without hardcoding would be to store a `PIN_CHANNEL_ID` constant in config somewhere, then add a command to set the pin channel, e.g. .set_pin_channel, which will basically execute `PIN_CHANNEL_ID = ctx.channel.id` and write it out to config. It's most likely overkill but it is robust and avoids any hardcoding which kinda gets me off.

Any of those options sound fine to me so I'll leave it up to you.

MonkfishFillet replied:
Def agree with setting a constant in there, I'm on it. I'll see if I can do a bit of digging to find the best solution to make it dynamic but realistically I think of a solution that fits all use-cases. if we search by name and the name is changed for whatever reason, broke. If we search by id and we delete it and make a new one, broke. If we delete it and change the name, double broke. Just the nature of the beast, knowwhatimsayinmatelol

For now, I think any dynamic solution is beyond the scope of this PR so I'll set a constant and come back to it later.

DonChalupa commented on cogs/Pinner.py:
```diff
@@ -0,0 +1,37 @@
+            for embed in pin.embed:
+                await pin_channel.send(embed=embed)
+
+            emoji = '<:bigfoot:468234675622641674>'
```
Prefer not to hardcode this here, for all the same reasons listed in the comment above, but also difficult to implement without hardcoding at all (also for the same reasons above). 

At the very least, could be moved to config as `PIN_REACTION` or something, so that if it does break it only needs to be fixed in one place and that place is clear. Could also add a command that sets this reaction if you wanted to go balls out.

MonkfishFillet replied:
Same as above, adding constant to Pinner Class

DonChalupa commented on models/Drink.py:
```diff
@@ -0,0 +1,50 @@
+        self._instructions = _json.get('strInstructions')
+        self.embed = embedder.embed(self)
+
+    def __get_ingredient_string(self, json) -> list:
```
Did you intend to use `__get_ingredient_string` here instead of `_get_ingredient_string` for the name mangling? Not saying there's a problem with it or it should be changed, just want to understand if there's a particular reason for the double underscores.

MonkfishFillet replied:
Really not much of reason - thought process was I saw no reason to keep it as a protected class if the model wasn't going to have children classes inherited from it. Making it private just seemed like it was a better fit

DonChalupa commented on services/Embedder.py:
```diff
@@ -0,0 +1,62 @@
+from abc import ABC, abstractmethod
+from discord import Embed
+from util.Utility import Colors
```
No longer exists after deletion of `util.Utility`.
```suggestion
from discord import Colour
```

MonkfishFillet replied:
Had to fiddle with this a bit - apparently there was a difference between `from discord import Colour` and `from discord,colour import Colour`.  Resolved now.

Additionally, `from discord,colour import Color` works as well - this is what I ran with because because

DonChalupa commented on services/Embedder.py:
```diff
@@ -0,0 +1,62 @@
+    @staticmethod
+    def embed(drink) -> Embed:
+
+        embed = Embed(title= drink._name, description='A drink for you, good Bud.', color=Colors().dark_navy)
```
```suggestion
        embed = Embed(title= drink._name, description='A drink for you, good Bud.', color=Colour.dark_blue())
```

DonChalupa commented on services/Embedder.py:
```diff
@@ -0,0 +1,62 @@
+
+    @staticmethod
+    def embed(pin) -> Embed:
+        embed = Embed(title=pin.author, description=f'Posted on {pin.posted_date}', color=int(Colors().user_colors.get_user_color(pin.id)))
```
`Colors()` no longer exists after `util.Utility` deletion

DonChalupa reviewed (changes requested):
Looked through stuff a little more. Still need to understand the interplay between cogs/models/services, that kind of paradigm isn't something I'm familiar with.

MonkfishFillet commented:
I lied, music isn't working due to missing packages and also some other stuff looool

Commit:
Utility Removal Cleanup

Commit:
Revert "Fix bug where message files weren't being saved to the data directory. Make sure the guild's data directory exists before trying to save there."

This reverts commit e002430f5deb6f3de1cd7608e08dca25a59fc635.

Commit:
Merge with Don Develop

Commit:
Added youtube-dl to requirements

Commit:
Removed Environment Folder

MonkfishFillet commented:
Fixed those problems. I removed robbot-env since we can replicate it using venv and the requirment.txt, and the files generated aren't necessary on the repo

MonkfishFillet committed:
Moved Services into Corresponding Cogs"

MonkfishFillet committed:
Refactored Embedder

MonkfishFillet committed:
Removed Unnecessary Config Values
